### ISSUE TITLE
Minor Implementation Questions

### ISSUE THREAD

**Issue by volkbay** [paulhager/MMCL-Tabular-Imaging#26]
Hi,

Thank you for your research efforts and for making them open-source. I would like to test your ideas by integrating them into our framework, yet I have encountered some minor issues that confused me:

1. In Section 4.2 of the publication (fetched from arxiv), it is stated that "_our tabular projector generates projections directly from
the embeddings with a fully connected layer._", but in  [pretraining.py](../blob/main/models/pretraining.py) file `SimCLRProjectionHead` is called as the tabular projector, which I am guessing is an MLP. Hence this is the case for the tabular pretraining models (MMCL, SCARF). Not a big issue, but which one is more suitable, MLP or FC?

2. I understand that the standard loss is [clip_loss.py](../blob/main/utils/clip_loss.py) for your proposed method. While it is formulated like below in the paper, it is neatly coded using `nn.CrossEntropyLoss`. Still, I don't think that `k≠j` condition is included in the softmax part of the CE loss. Am I wrong with it? Or do you think this is still valid?

![image](https://github.com/user-attachments/assets/b918e9af-bb06-4768-a331-8e4d97c6c505)

3. This one is ineffective in training, so I just wanted to let you know. There is a `ReLU` function at the beginning of the tabular projector in [TabularEmbeddingModel.py](../blob/main/models/TabularEmbeddingModel.py), which seem like the only instance there is an activation between encoders and projectors. This may not yield the same results if you have decided to keep the projector after pretraining.

4. For those that are new to this framework and looking up the issues, `lr_imaging`, `lr_tabular` and `lr_classifier` among configurations seem redundant and a bit confusing in the beginning. They are possibly for a test case or obselete.

I hope I am clear/correct with those and will not take your time. Great work, thanks!

**Comment by paulhager** [comment]
1. We started with a FC initially but found the one layer MLP to be a bit more stable. Our experiments didn't show much of a difference in final performance though. I'd recommend using the one layer MLP as in the code. Good catch though, this is imprecise in the paper.
2. Our implementation follows that of the pseudocode of the original CLIP paper and the OpenCLIP repository which simply calculates CrossEntropy between logits. Our math follows the text of the original CLIP paper which states that ".. maximize the cosine similarity of the image and text embeddings of the N real pairs in the batch while minimizing the cosine similarity of the embeddings of the N^2 − N incorrect pairings." It would be cleaner to write the math without the k != j condition to reflect the implementation.
3. You are completely correct here, this is an error. I've fixed this and pushed it. Thanks for finding this.
4. Yes, I had experimented with different learning rates for the two modalities but found a single learning rate to be most robust and performant. I've removed it from the YAML.

Thanks for your close reading of the repository! If you find anything else let me know.
